tests/db/utils_test.py

```diff
@@ -380,3 +387,235 @@ def test_unicode_as_bytes(self):
+            or according to the mechanism defined in that document by
+            converting the text name to lower case and prefixing it with the
+            four characters "pgp-".
+
+        The spec is pretty clear that this is supposed to be lower cased.
+        """
+        m = self._make_signed()
+        m.set_param('micalg', 'PGP-SHA1')
+        m = utils.message_from_file(io.BytesIO(m.as_string()))
+        self.assertIn('expected micalg=pgp-',
+                      m[utils.X_SIGNATURE_MESSAGE_HEADER])
+
+    def test_signed_more_than_two_messages(self):
+        """Per the spec only 2 payloads may be encapsulated inside the
+        multipart/signed payload, while it might be nice to cover more than 2
+        payloads (Postel's law), it would introduce serious complexity
+        since we would also need to cover those payloads being misordered.
+        Since getting the right number of payloads and getting them in the
+        right order should be fairly easy to implement correctly enforcing that
+        there are only two payloads seems reasonable.
+        """
+        m = self._make_signed()
+        m.attach(email.mime.text.MIMEText('foo'))
+        m = utils.message_from_file(io.BytesIO(m.as_string()))
+        self.assertIn('expected exactly two messages, got 3',
+                      m[utils.X_SIGNATURE_MESSAGE_HEADER])
+
+    # TODO: The case of more than two payloads, or the payloads being out of
+    # order. Also for the encrypted case.
+
+    def _make_encrypted(self, signed=False):
+        """Create an encrypted (and optionally signed) message."""
+        if signed:
+            t = self._make_signed()
+        else:
+            text = 'This is some text'
+            t = email.mime.text.MIMEText(text, 'plain', 'utf-8')
+        enc = crypto.encrypt(t.as_string(), keys=[crypto.get_key('F74091D4133F87D56B5D343C1974EC55FBC2D660')])
+        e = email.mime.application.MIMEApplication(
+            enc, 'octet-stream', email.encoders.encode_7or8bit)
+
+        f = email.mime.application.MIMEApplication(
+            b'Version: 1', 'pgp-encrypted', email.encoders.encode_7or8bit)
+
+        m = email.mime.multipart.MIMEMultipart('encrypted', None, [f, e])
+        m.set_param('protocol', 'application/pgp-encrypted')
+
+        return m
+
+    def test_encrypted_length(self):
+        # It seems strang that we just attache the unsigned message to the end
+        # of the mail, rather than replacing the whole encrypted payload with
+        # it's unencrupted enquivalent
+        m = self._make_encrypted()
+        m = utils.message_from_file(io.BytesIO(m.as_string()))
+        self.assertEqual(len(m.get_payload()), 3)
+
+    def test_encrypted_unsigned_is_decrypted(self):
+        m = self._make_encrypted()
+        m = utils.message_from_file(io.BytesIO(m.as_string()))
+        self.assertEqual(m.get_payload(2).get_payload(), 'This is some text')
```

Comment:
Is there some more explicit way to check that the message is not signed? This relies on the knowlesge how a unsigned encrypted mime message has to be structured.